### PR TITLE
Re-write of colour fix code

### LINKED ISSUE #19: wrong colour when second model plots into empty axes
When plotting 2 models one of which has more parameters than the other, e.g. LCDM and rLCDM, and if we then plot the model with less parameters first including the unshared parameter, then the second model will have the wrong colour in the axes with the unshared parameter.

The colour should be dependent on the call to `plot_2d` and thus on the model. It should not depend on whether any axis is drawn in for the first or subsequent time.

Example code:
```python
fig, axes= ns_lcdm.plot_2d(['A_s', 'n_s', 'r'], label='LCDM')
ns_rlcdm.plot_2d(axes=axes, diagonal_kwargs={'ls': '--'}, label='rLCDM')
leg = get_legend_proxy(fig)
axes['n_s']['n_s'].legend(leg, ['LCDM', 'rLCDM'], bbox_to_anchor=(1, 1), loc='lower left')
```

image

Note, how in the above code the second model gets the additional keyword `ls='--'` which shows that rLCDM is drawn in _orange_ in the upper two axes rows and _blue_ in the bottom row.

For now this behaviour can be circumvented by plotting the model with the most parameters first. This won't help in cases where two models each have an additional parameter not shared with the other, though, say e.g. r+LCDM and alpha+LCDM.

### PR DESCRIPTION
**Pull request by williamjameshandley**
@lukashergt: is there any reason why this wouldn't fix #19 in the same manner as #15?

I think this way of doing it is slightly more logical, and will be easier for downstream changes to integrate.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code is PEP8 compliant (`flake8 anesthetic tests`)
- [x] My code contains compliant docstrings (`pydocstyle --convention=numpy anesthetic`)
- [x] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: anesthetic/utils.py [hunk 1]
```diff
@@ -55,10 +55,11 @@ def quantile(a, q, w):
...
     """Check if we need to apply strict bounds."""
-    if xmin is not None and (d.min() - xmin)/(d.max()-d.min()) > 1e-2:
-        xmin = None
-    if xmax is not None and (xmax - d.max())/(d.max()-d.min()) > 1e-2:
-        xmax = None
+    if len(d) > 0:
+        if xmin is not None and (min(d) - xmin)/(max(d)-min(d)) > 1e-2:
+            xmin = None
+        if xmax is not None and (xmax - max(d))/(max(d)-min(d)) > 1e-2:
+            xmax = None
```
Comment (lukashergt) [inline anesthetic/utils.py]: I'm getting this error now:
```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-26-3db102adf145> in <module>()
----> 1 fig, axes= pc_lcdm.plot_2d(['A_s', 'n_s', 'r'], label='LCDM')

/scratch/.virtualenvs/py27env/lib/python2.7/site-packages/anesthetic/samples.pyc in plot_2d(self, axes, *args, **kwargs)
    278                     ax_ = ax.twin if x == y else ax
    279                     self.plot(ax_, x, y, plot_type=types[pos], *args,
--> 280                               **local_kwargs[pos])
    281 
    282         return fig, axes

/scratch/.virtualenvs/py27env/lib/python2.7/site-packages/anesthetic/samples.pyc in plot(self, ax, paramname_x, paramname_y, *args, **kwargs)
    155 
    156             return plot(ax, x, y, xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax,
--> 157                         *args, **kwargs)
    158 
    159     def plot_1d(self, axes, *args, **kwargs):

/scratch/.virtualenvs/py27env/lib/python2.7/site-packages/anesthetic/plot.pyc in contour_plot_2d(ax, data_x, data_y, *args, **kwargs)
    351 
    352     x, y, pdf = kde_2d(data_x, data_y,
--> 353                        xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax)
    354     pdf /= pdf.max()
    355     p = sorted(pdf.flatten())

/scratch/.virtualenvs/py27env/lib/python2.7/site-packages/anesthetic/kde.pyc in kde_2d(d_x, d_y, xmin, xmax, ymin, ymax)
     79     f = [xmax is None or xmin is None,
     80          ymax is None or ymin is None]
---> 81     d_x_, d_y_ = mirror_2d(d_x, d_y, xmin, xmax, ymin, ymax)
     82 
     83     with warnings.catch_warnings():

/scratch/.virtualenvs/py27env/lib/python2.7/site-packages/anesthetic/utils.pyc in mirror_2d(d_x_, d_y_, xmin, xmax, ymin, ymax)
     79 def mirror_2d(d_x_, d_y_, xmin=None, xmax=None, ymin=None, ymax=None):
     80     """If necessary apply reflecting boundary conditions."""
---> 81     d_x = d_x_.copy()
     82     d_y = d_y_.copy()
     83 

AttributeError: 'list' object has no attribute 'copy'
```

Comment (williamjameshandley) [reply anesthetic/utils.py]: How easy would it be to add a failing test which shows this error?

Comment (lukashergt) [reply anesthetic/utils.py]: Test added.

The test for `samples.plot_2d` didn't check for 1d lists as input.

**Comment by williamjameshandley** [comment]
@lukashergt I believe this is now fixed, although would be good to check that you don't get any further issues arising on your system.

**Pull request merged**